synapse/util/caches/descriptors.py

```diff
@@ -182,16 +188,28 @@ def __get__(self, obj, objtype=None):
             name=self.orig.__name__,
             max_entries=self.max_entries,
             keylen=self.num_args,
-            lru=self.lru,
             tree=self.tree,
         )
 
         @functools.wraps(self.orig)
         def wrapped(*args, **kwargs):
+            cache_context = kwargs.pop("cache_context", None)
+            if cache_context:
+                context_callback = cache_context.invalidate
+            else:
+                context_callback = None
+
+            self_context = _CacheContext(cache, None)
+            if self.add_cache_context:
+                kwargs["cache_context"] = self_context
```

Comment:
I feel its nicer to have a cache context style thing, as its far from implausible we may want to expose other functionality/information from the cache in it.